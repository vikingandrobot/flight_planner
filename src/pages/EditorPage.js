import React from 'react';

import styles from './EditorPage.module.scss';

function EditorPage() {
  return (
    <div className={styles.editorPage}>
      <h1>Flight Planner</h1>
    </div>
  );
}

export default EditorPage;
