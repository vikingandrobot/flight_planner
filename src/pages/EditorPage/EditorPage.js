import React, { useState, useEffect } from 'react';
import { Button, Input, Typography } from 'antd';
import { PlusCircleOutlined, CheckCircleOutlined } from '@ant-design/icons';
import { fromLonLat } from 'ol/proj';

import { get, post } from '../../api/FlightPlansAPI';
import Map from '../../components/Map/Map';
import FlightPlanList from '../../components/FlightPlanList/FlightPlanList';

import styles from './EditorPage.module.scss';

const { Title } = Typography;

const LONGITUDE = 6.5685000;
const LATITUDE = 46.5185000;

const center = fromLonLat([LONGITUDE, LATITUDE]);
const zoom = 17;

function NewFlightActions({ onCreate, onCancel, disabled }) {
  return (
    <>
      <Button
        type="primary"
        onClick={onCreate}
        disabled={disabled}
        icon={<CheckCircleOutlined />}
      >
        Create Flight Plan
      </Button>
      <Button
        onClick={onCancel}
      >
        Cancel
      </Button>
    </>
  )
}

function DefaultActions({ onNew }) {
  return (
    <Button
      type="primary"
      icon={<PlusCircleOutlined />}
      onClick={onNew}
    >
      New Flight Plan
    </Button>
  )
}

function EditorPage() {
  const [isLoadingData, setIsLoadingData] = useState(false);
  const [isCreating, setIsCreating] = useState(false);
  const [flightPlans, setFlightPlans] = useState([]);
  const [selectedFlightPlanId, setSelectedFlightPlanId] = useState(null);
  const [draftFlightPlan, setDraftFlightPlan] = useState(null);

  useEffect(() => {
    setIsLoadingData(true);
    get()
      .then(setFlightPlans)
      .finally(() => setIsLoadingData(false));
  }, []);

  function handleNew() {
    setIsCreating(true);
    setSelectedFlightPlanId(null);
    setDraftFlightPlan({ title: null, points: [] });
  }

  function handleCreate() {
    setIsLoadingData(true);
    post(draftFlightPlan)
      .then(get)
      .then((plans) => {
        setDraftFlightPlan(null);
        setIsCreating(false);
        setFlightPlans(plans);

      })
      .finally(() => {
        setIsLoadingData(false);
      });
  }

  function handleCancel() {
    setIsCreating(false);
    setDraftFlightPlan(null);
  }

  function handleSelect({ id }) {
    setSelectedFlightPlanId(id);
  }

  function handleTitleChange(e) {
    setDraftFlightPlan(prevDraft => ({ ...prevDraft, title: e.target.value }));
  }

  function handleDraftClick(coordinate) {
    setDraftFlightPlan(prevDraft => ({ ...prevDraft, points: [...draftFlightPlan.points, coordinate] }));
  }

  return (
    <div className={styles.editorPage}>
      <header className={styles.pageHeader}>
        <Title level={1}>Flight Planner</Title>
      </header>
      <section className={styles.editorContent}>
        <header className={styles.editorHeader}>
          <div>
            {isCreating && (
              <Input
                placeholder="Flight Plan name"
                onChange={handleTitleChange}
              />
            )}
          </div>

          <div className={styles.actionsContainer}>
            {isCreating ?
            <NewFlightActions
              onCreate={handleCreate}
              onCancel={handleCancel}
              disabled={!draftFlightPlan || draftFlightPlan.points.length < 2}
            /> :
            <DefaultActions onNew={handleNew} />}
          </div>
        </header>
        <div className={styles.editorMapContainer}>
          <Map
            center={center}
            className={styles.editorMap}
            fitViewToFlightPlan={!isCreating}
            flightPlan={isCreating ? draftFlightPlan : (selectedFlightPlanId && flightPlans.find(fp => fp.id === selectedFlightPlanId))}
            onClick={isCreating ? handleDraftClick : null}
            zoom={zoom}
          />
          <FlightPlanList
            className={styles.editorFlightList}
            flightPlans={flightPlans}
            loading={isLoadingData}
            onRowClick={handleSelect}
          />
        </div>
      </section>
    </div>
  );
}

export default EditorPage;
