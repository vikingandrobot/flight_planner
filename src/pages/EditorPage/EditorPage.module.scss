@import '../../styles/colors.scss';

.editorPage {
  background-color: $veryLightBackground;
  width: 100vw;
  height: 100vh;
  color: $darkTextColor;
  display: flex;
  flex-direction: column;

  .pageHeader {
    padding: 5px 10px;
    background: $darkBlueBackground;
    height: auto;
    h1 {
      color: $veryLightTextColor;
      margin: 0;
    }
  }

  .editorContent {
    position: relative;
    flex-grow: 1;
    display: flex;
    flex-direction: column;

    .editorHeader {
      display: flex;
      position: relative;
      padding: 5px 10px;
      background-color: $veryLightBackground;

      & > div {
        flex-grow: 1;
      }
    }

    .actionsContainer {
      display: flex;
      flex-direction: row-reverse;
    }

    .editorMap {
      flex-grow: 1;
      width: 100%;
      height: 100%;
    }
    .editorFlightList {
      height: 100%;
    }
  }
}
