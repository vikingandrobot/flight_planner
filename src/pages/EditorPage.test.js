import { render, screen } from '@testing-library/react';

import EditorPage from './EditorPage';

describe('EditorPage', () => {
  test('should render the editor page', () => {
    render(<EditorPage />);
    const linkElement = screen.getByText('Flight Planner');
    expect(linkElement).toBeInTheDocument();
  })
})
