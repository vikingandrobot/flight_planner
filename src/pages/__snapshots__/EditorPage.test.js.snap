// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage should render the editor page 1`] = `
<div
  className="editorPage"
>
  <header>
    <h1>
      Flight Planner
    </h1>
  </header>
  <section
    className="editorMapContainer"
  >
    <Map
      className="editorMap"
      viewOptions={
        Object {
          "center": Array [
            731202.0752756175,
            5863832.126542759,
          ],
          "zoom": 17,
        }
      }
    />
  </section>
</div>
`;
